Boost: Save concatenated JS/CSS in files on server

dilirity opened the pull request:
Related https://github.com/Automattic/jetpack/issues/41035

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
*

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Apply this PR
* Make sure the two concatenate modules are enabled in Boost settings.
* Load up your site and confirm that the minified CSS/JS is loaded from wp-content/boost-cache/static/.
* If not, then toggle one of the concatenate modules.
* If the minified CSS/JS is still being served from "/_jb_static/??" then mod_rewrite is disabled in wp-content/
* Examine your site .htaccess and make sure there's nothing in there that stops it working in that directory.
* Look for wp-content/.htaccess and make sure it doesn't contain `RewriteEngine Off`.

To test the is_404() tester:
* Make sure the minified files are served from wp-content/boost-cache/static/.
* Add wp-content/.htaccess with the single line `RewriteEngine Off`.
* Toggle one of the concatenate modules.
* Reload your blog.
* The Page Cache isn't deleted by this, the site owner will have to clear the cache.

Test the cache cleanup when either concatenate module is toggled off
* Make sure there are static files in wp-content/boost-cache/static/
* Toggle either concatenate module off.
* Make sure the corresponding file is deleted.
* Repeat for the other module.

dilirity committed:
Update URLs for concatenated assets

dilirity committed:
Add WIP code to handle new URL

dilirity committed:
add changelog

dilirity committed:
Fix URLs for concatenated assets pointing to outside wp-content files

dilirity committed:
Store concatenated output in files

dilirity committed:
Update file name to indicate that file is minified

dilirity committed:
Refactor serving and storing of concatenated files

dilirity committed:
Refactor file name building

dilirity committed:
Pass request uri instead of reaching in global array

dilirity committed:
Rely on existing code for allowed extensions instead of hard-coded list

dilirity committed:
Move handling static concatenation to a separate file

donnchawp committed:
Enhance cache cleanup functionality.

* Files are now purged based on last access time, one week.
* A separate cleanup for static cache directories.
* Files older than a week are removed regardless of access time.

donnchawp committed:
Don't deviate too much from how optimize used to garbage collect.

Instead of a week, go with twice the access time, which is 2 days.
Longer than that, and we will need to rebuild the cached files in
a cron job, and then we won't know how many cached minified files
were really obsolete and shouldn't be updated any more

donnchawp committed:
Load the minification engine that builds the minified file

donnchawp committed:
Load the minification loader

donnchawp committed:
Always load minification classes, for maintenance tasks

donnchawp committed:
Allow new static cache file to go for a week

The old files exist for only a day because they're not checked.
I'll add a cron job that tries to maintain the new files.

donnchawp committed:
loader.php has to be loaded from the template_redirect action

donnchawp committed:
Add a maintenance function that deletes obsolete minified files

This function compares the minified file with the dependencies. If they
are newer, then the minified file is deleted, so it can be regenerated
by the 404 handler.

donnchawp committed:
/Users/donncha/dev/jetpack/vendor/sirbrillig/phpcs-Fixes for PHAN errors.

projects/plugins/boost/app/lib/minify/functions-service-new.php:7
Error: Plugin PhanPluginNeverReturnFunction Function \jetpack_boost_handle_minify_request() never returns and has a return type of void, but phpdoc type never could be used instead
projects/plugins/boost/app/lib/minify/functions-service-new.php:156
Error: Plugin PhanPluginUseReturnValueInternalKnown Expected to use the return value of the internal function/method \preg_replace_callback
projects/plugins/boost/app/lib/minify/functions-service-new.php:213
Error: UndefError PhanPossiblyUndeclaredVariable Variable $mime_type is possibly undeclared
projects/plugins/boost/app/lib/minify/functions-service-new.php:243
Error: TypeError PhanTypePossiblyInvalidDimOffset Possibly invalid offset 'extension' of $file_info of array type array{dirname:string,basename:string,extension?:string,filename:string}
projects/plugins/boost/app/lib/minify/functions-service-new.php:254
Error: TypeError PhanTypePossiblyInvalidDimOffset Possibly invalid offset 'extension' of $file_info of array type array{dirname:string,basename:string,extension?:string,filename:string}

donnchawp committed:
Always load minification include files to avoid problems uninstalling

jetpack_boost_cache_maintenance() in functions-service-new.php is used
by the helper script when uninstalling, but won't be loaded when
uninstalling if it's loaded on template_redirect.

donnchawp committed:
This file is loaded by app/lib/minify/loader.php now

donnchawp committed:
If file_age is 0 then delete files.

* No need to maintain obsolete static files. Just delete them.
* Make sure to delete static files in boost-cache when file_age is 0
* No need to check filemtime or fileatime. Just delete.

donnchawp committed:
Add comment to explain jetpack_boost_cache_maintenance()

donnchawp committed:
Just in case WP is installed in a directory. Not always at "/"

donnchawp commented:
There's an issue with multisite WordPress installs using domains. With the default mod_rewrite rules described [here](https://developer.wordpress.org/advanced-administration/server/web-server/httpd/#multisite) any missing file in wp-content/ causes an internal server error. The solution is to modify the rules. Duplicate the 2 RewriteCond conditions here:

```
RewriteCond %{REQUEST_FILENAME} -f [OR]
RewriteCond %{REQUEST_FILENAME} -d
RewriteRule ^(wp-(content|admin|includes).*) $1 [L]
```
Also, there are probably WordPress sites out there that force everything in wp-content to be served by the web server, not PHP. If that's so, requesting a missing file in that directory will show a regular 404 page, not the WordPress one.

We possibly need an is_404 check when either the concatenation modules is toggled on. It would request a missing URL in wp-content, and if the is_404 check is activated and the URL matches, set a transient or some flag. The calling code then checks for that flag to verify that is_404() works. If not, we fall back to the old way of serving cached minified files.

donnchawp committed:
Fix the maintenance script so it gets the cached list of files

donnchawp committed:
Only check moved cache folder for legacy folder

donnchawp committed:
Add is_404() tester to minification

donnchawp committed:
Are we testing the is_404() handler?

donnchawp committed:
Test the is_404() check when activating minification modules

donnchawp committed:
Serve legacy minified files if new static files not supported

donnchawp committed:
Fix the @charset bug in the old minified files too

donnchawp committed:
Run the 404 tester at least once in wp-admin/

donnchawp committed:
Updated PHAN baseline

donnchawp committed:
Updated PHAN baseline

donnchawp committed:
Use str_contains instead of strpos

Trying to force the e2e tests to run again

donnchawp committed:
Store static minification bool in a site option

On a multisite install, this will stop the plugin checking if is_404
works every time a new user logs in to /wp-admin/.

donnchawp committed:
Add code to cleanup cached files of deactivated module

It will delete the corresponding static files, css or js.

LiamSarsfield commented on projects/plugins/boost/app/class-jetpack-boost.php:
```diff
@@ -181,7 +181,6 @@ public function deactivate() {
 		do_action( 'jetpack_boost_deactivate' );
 
 		// Tell Minify JS/CSS to clean up.
-		require_once JETPACK_BOOST_DIR_PATH . '/app/lib/minify/functions-helpers.php';
```
Just confirming: this is removed because it is now being autoloaded in `projects/plugins/boost/app/lib/minify/loader.php`?

donnchawp replied:
Yeah, that's it.

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/Concatenate_CSS.php:
```diff
@@ -195,9 +195,13 @@ public function do_items( $handles = false, $group = false ) {
+					if ( get_site_option( 'jetpack_boost_static_minification' ) ) {
+						$href = jetpack_boost_get_minify_url( $file_name . '.min.css', $siteurl );
+					} else {
+						$href = $siteurl . jetpack_boost_get_static_prefix() . '??' . $file_name;
```
Just curious - what does the `'??'` mean in this context?

donnchawp replied:
The old way of delivering the concatenated files was using the URL /_jb_static/??[hash]. It's now the fallback. If you look at your test site using trunk you'll see that for css and js files if you have the concatenation module enabled.

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/Config.php:
```diff
@@ -28,13 +32,47 @@ public static function get_abspath(): string {
 
+	public static function can_use_static_cache() {
+		$cache_dir = static::get_static_cache_dir_path();
+		$use_cache = ! empty( $cache_dir );
```
Nitpick: `$use_cache` is being checked multiple times below and we are returning it if it's false anyways.
Could we return early if we see that `$use_cache` is false on each check? That way we wouldn't need to keep checking that it's true, as it's going to be true if the code reaches that point.

donnchawp replied:
Good catch. Peter copied the code from can_use_cache() which has these extra checks because get_cache_dir_path() can return false.
get_static_cache_dir_path() only returns the path, so no need for $use_cache at all.

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/Config.php:
```diff
@@ -28,13 +32,47 @@ public static function get_abspath(): string {
+
 	public static function can_use_cache() {
 		$cache_dir = static::get_cache_dir_path();
 		$use_cache = ! empty( $cache_dir );
```
We could use a similar pattern as [mentioned here](https://github.com/Automattic/jetpack/pull/41056/files#r1935451145) too, but I see that this could be out of scope as there is code that might need to be slightly refactored in this function that wasn't changed.

donnchawp replied:
I decided to change it and add a $use_cache check there which will log the error in the same way as the other checks. Thanks!

donnchawp committed:
The static cache dir doesn't change, it's a string always

We don't as yet allow the static cache dir to be changed so it's safe to
remove the $use_cache checks. get_static_cache_dir_path() always returns
a path to the directory.

donnchawp committed:
Return early if $cache_dir is not defined.

No need to go through other checks. It will log a proper error for this
too.

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/Config.php:
```diff
@@ -28,13 +32,57 @@ public static function get_abspath(): string {
+						$cache_dir
+					)
+				);
+			}
```
Nit: I was thinking we could just return false here, unless we're okay with potentially having 2 logs when the directory does not exist? As I think "Disabling concatenate static cache. Unable to create cache directory" and "Disabling concatenate static cache. Unable to write to cache directory" could be called twice.

LiamSarsfield replied:
Actually scratch that, I see that `ensure_dir_exists` also makes the directory if it doesn't exist.

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/loader.php:
```diff
@@ -0,0 +1,32 @@
+
+		// Get the path for the current request.
+		$request_uri = remove_query_arg( 'JB_NONEXISTENTQUERY_ARG' );
+		if ( ! str_contains( strtolower( $request_uri ), '/wp-content/boost-cache/static/' ) ) {
```
[str_contains](https://www.php.net/manual/en/function.str-contains.php) is only supported for PHP 8.0+ but from looking at `projects/plugins/boost/jetpack-boost.php` we support 7.2+

dilirity replied:
Thanks for the question! Had to look things up!

WordPress has pollyfils for newer functions. These have been introduced in WP 5.9 and since we support [6.6+](https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/boost/jetpack-boost.php#L19) it's safe for us to use them.

[Here's the reference](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/compat.php#L484-L504).

LiamSarsfield replied:
Oh nice! Good to know, this will be very handy.

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/functions-service-new.php:
```diff
@@ -0,0 +1,348 @@
+ * Using a crafted request, we can check if is_404() is working in wp-content/
+ */
+function jetpack_boost_check_404_handler( $request_uri ) {
+	if ( ! str_contains( strtolower( $request_uri ), 'wp-content/boost-cache/static/testing_404' ) ) {
```
Just curious, why is this specifically checked `wp-content/boost-cache/static/testing_404 `?

LiamSarsfield replied:
Also another example of `str_contains` that I think we should remove if we support PHP versions < 8

dilirity replied:
See https://github.com/Automattic/jetpack/pull/41056#discussion_r1940871793

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/functions-service-new.php:
```diff
@@ -0,0 +1,348 @@
+ * Using a crafted request, we can check if is_404() is working in wp-content/
+ */
+function jetpack_boost_check_404_handler( $request_uri ) {
+	if ( ! str_contains( strtolower( $request_uri ), 'wp-content/boost-cache/static/testing_404' ) ) {
```
Just curious, why is this specifically checked `wp-content/boost-cache/static/testing_404 `?

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/Config.php:
```diff
@@ -66,4 +113,22 @@ public static function can_use_cache() {
+		return true;
+	}
+
+	private static function ensure_dir_is_writable( string $dir ): bool {
```
Nitpick: I originally thought this function would do an action (make the directory writable) like ensure_dir_exists does with it's action (make the directory).
I wonder if there's a better name for this 🤔 e.g. `is_dir_writable`

dilirity replied:
Makes more sense. Updated.

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/Config.php:
```diff
@@ -28,13 +32,57 @@ public static function get_abspath(): string {
+
+		if ( ! static::ensure_dir_exists( $cache_dir ) ) {
+			$use_cache = false;
+			if ( defined( 'WP_DEBUG' ) && WP_DEBUG ) {
```
For future consideration: I noticed this pattern is done throughout the code multiple times where we check `WP_DEBUG` and log an error if it's true.
I wonder if we can make this less repeatable, I think this is out of scope for this PR, but food for thought perhaps.

dilirity replied:
Thanks! I've addressed this in 686bd0e6f4ef5a4aa11eed975a6521ea168d4ecf.

I've also cleaned up the methods in the class to use early returns and include docblocks.

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/functions-service-new.php:
```diff
@@ -0,0 +1,348 @@
+ * It removes files that are stale and no longer needed.
+ * A file is considered stale if it's older than the files it depends on.
+ */
+function jetpack_boost_cache_maintenance() {
```
Nitpick: Would `jetpack_boost_minify_clean_stale_files` be a more accurate name than `jetpack_boost_cache_maintenance`? I was originally thinking the maintenance would do more actionable things than deleting the stale files.

dilirity replied:
Renamed for clarity.

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/functions-helpers.php:
```diff
@@ -31,21 +41,33 @@ function jetpack_boost_page_optimize_cache_cleanup( $cache_folder = false, $file
 	// Grab all files in the cache directory
-	$cache_files = glob( $cache_folder . '/page-optimize-cache-*' );
+	if ( $cache_folder === Config::get_static_cache_dir_path() ) {
+		$cache_files = glob( $cache_folder . '/*min*' );
```
Is there a reason why we check `glob( $cache_folder . '/*min*' )` here, but in `jetpack_boost_cache_maintenance` we use `glob( Config::get_static_cache_dir_path() . '/*.min.*' )`?

donnchawp replied:
Good catch. There's no reason why. The files are in the form [hash].min.[js|css] so either pattern will match.
I'll change the \*min\* one to \*.min.\* as it's more precise.

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/functions-helpers.php:
```diff
@@ -13,12 +13,22 @@ function jetpack_boost_minify_cache_buster() {
  * @param string $cache_folder The path to the cache folder to cleanup.
  * @param int    $file_age The age of files to purge, in seconds.
  */
 function jetpack_boost_page_optimize_cache_cleanup( $cache_folder = false, $file_age = DAY_IN_SECONDS ) {
```
Nitpick: We could actually name this function `jetpack_boost_cache_maintenance` instead as it seems this is actually doing the entire maintainance of the page optimize.

donnchawp replied:
That's a good point. I'll go one step further and say that I should rename the maintenance function `jetpack_boost_static_cache_cleanup` since that's the only cache it operates on. "maintenance" isn't descriptive enough. It could imply that it rebuilds the cache file, which it doesn't do.

donnchawp replied:
Peter already renamed it to jetpack_boost_minify_remove_stale_static_files().

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/functions-helpers.php:
```diff
@@ -273,6 +295,10 @@ function jetpack_boost_get_static_prefix() {
 	return trailingslashit( $prefix );
 }
 
+function jetpack_boost_get_minify_url( $file_name, $siteurl ) {
```
Should `$siteurl` be `$site_url`?

dilirity replied:
Yep. Fixed.

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/functions-service-new.php:
```diff
@@ -0,0 +1,348 @@
+	}
+
+	$file_info = pathinfo( $file_path );
+	if ( stripos( $file_info['dirname'], '/wp-content/boost-cache/static' ) === false ) {
```
What does this do? Would a comment here help for readability?

dilirity replied:
Updated in 0da903c6295faca61a2738c0d87888b3602303ab. Not sure if the `realpath` checks are necessary. Just wanted to be safe 😅 

@donnchawp - what do you think of this? Is it okay for MU sites?

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/functions-service-new.php:
```diff
@@ -0,0 +1,348 @@
+	$use_cache = Config::can_use_static_cache();
+	if ( $use_cache ) {
+		$file_parts = jetpack_boost_minify_get_file_parts( $request_uri );
+		if ( $file_parts ) {
```
Should we either check if this is an array, and/or explicitly check the array has the array keys `file_name` and `file_extension`?

dilirity replied:
It does now, thanks 😅

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/functions-service-new.php:
```diff
@@ -0,0 +1,348 @@
+		wp_delete_file( Config::get_static_cache_dir_path() . '/404' );
+		update_site_option( 'jetpack_boost_static_minification', 1 );
+	} else {
+		update_site_option( 'jetpack_boost_static_minification', 0 );
```
This may be out of scope for this iteration, but we could have a notice within the Boost settings that communicates to the user they are not taking full advantage of our Concatenate JS/CSS

donnchawp replied:
I totally agree. There is a support doc describing the problem where we could point people, but we should probably leave linking to it to another PR. :)

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/functions-service-new.php:
```diff
@@ -0,0 +1,348 @@
+function jetpack_boost_404_setup() {
+	if ( is_admin() && get_site_option( 'jetpack_boost_static_minification', 'na' ) === 'na' ) {
+		update_site_option( 'jetpack_boost_static_minification', 0 ); // Add a default value if not set to avoid an extra SQL query.
+		wp_schedule_single_event( time() + 3, 'jetpack_boost_404_tester_cron' );
```
Just curious, what is the reasoning behind having this in a cron, and the benefit of having it be done in 3 seconds?

donnchawp replied:
I should probably expand on the comment there to explain this. It's an effort to stop a race condition. It's only possible to schedule a job once, so that ensures we only run the tester once.

The update_site_option() ensures that that particular chunk of code only runs as many times as however many users hit wp-admin/ simultaneously (well, plus the time MySQL takes to do the update).

I schedule the job for 3 seconds because if it was scheduled immediately, I'm paranoid that another process might run the previous `update_site_option()` at the same time and overwrite a successful test. If it's 3 seconds later, there's less chance of that happening.

LiamSarsfield commented on projects/plugins/boost/app/modules/Modules_Setup.php:
```diff
@@ -152,5 +154,9 @@ public function on_module_status_update( $module_slug, $is_activated ) {
 			( new Regenerate() )->start();
 		}
+
+		if ( $module_slug === Minify_CSS::get_slug() || $module_slug === Minify_JS::get_slug() ) {
```
We could leverage the Has_Activate/Has_Deactivate interface within `projects/plugins/boost/app/modules/Module.php` in order to call `jetpack_boost_static_minify_setup` within `projects/plugins/boost/app/modules/optimizations/minify/class-minify-css.php` and `projects/plugins/boost/app/modules/optimizations/minify/class-minify-js.php`
E.g. see below for example of `deactivate`
`projects/plugins/boost/app/modules/optimizations/page-cache/Page_Cache.php`

dilirity replied:
@donnchawp I've handled this in 9ceaacf. I also forgot we had such interfaces 😅

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/functions-service-new.php:
```diff
@@ -0,0 +1,348 @@
+ * @param bool   $activated Whether the module is activated.
+ * @param string $module_slug The module slug.
+ */
+function jetpack_boost_static_minify_setup( $activated, $module_slug ) {
```
Nitpick: This function does more than setup as it also cleans up when it is deactivated, we could have this logic within an implementation as mentioned here https://github.com/Automattic/jetpack/pull/41056#discussion_r1937111221

dilirity replied:
Function no longer exists as of changes introduced in 9ceaacf.

LiamSarsfield commented on projects/plugins/boost/app/lib/minify/functions-service-new.php:
```diff
@@ -0,0 +1,348 @@
+	$file_paths = jetpack_boost_page_optimize_get_file_paths( $file_parts['file_name'] );
+
+	// file_paths contain something like array( '/foo/bar.css', '/foo1/bar/baz.css' )
+	// @todo - what do we do if we're over the maximum number of files? Maybe we just don't concatenate?
```
Should this todo be addressed in this PR? If not, should we create an issue as a follow up?

dilirity replied:
Thanks.

Removed the todo as there's already an issue for this - https://github.com/Automattic/jetpack/issues/40365

LiamSarsfield reviewed (commented):
Works as expected per testing instructions! I do have a couple of comments after reviewing, mainly nitpicks, clarifications.

dilirity committed:
Remove todo

There's already an issue for that.

dilirity committed:
Improve readability of concat Config class

dilirity committed:
Rename helper method

dilirity committed:
Rename function

dilirity committed:
Fix incorrect return type

dilirity committed:
Rename files so they make more sense

dilirity committed:
Add proper check for file parts

dilirity committed:
Merge branch 'trunk' into update/boost/store-concat-in-files

dilirity committed:
Remove types

dilirity committed:
Update comments

dilirity committed:
Fix variable name

dilirity committed:
Update minify css/js to use activate/deactivate to run their setup

dilirity committed:
Fix fatal errors due to file renames

dilirity committed:
Fix incorrect json_decode usage

dilirity committed:
Fix static analysis complaining about possibly undefined variables

dilirity committed:
Fix incorrect comparison

dilirity committed:
Remove redundant check

dilirity committed:
Make phan happy

dilirity committed:
Update static analysis baseline

dilirity committed:
Improve path checking and add clarification

donnchawp committed:
Make the glob pattern more precise

donnchawp committed:
Merge branch 'update/boost/store-concat-in-files' of github.com:Automattic/jetpack into update/boost/store-concat-in-files

donnchawp committed:
Fix the directory security check so static files can be created

donnchawp committed:
Use str_contains() to look for the string

It's checking for the existence of the string, not position.

donnchawp committed:
Avoid checking again if the file is deleted.

If the first check succeeded, the file will be deleted. No need to check
the filemtime() on a deleted file and cause a warning.

donnchawp committed:
Add a comment to describe what happens here.

donnchawp commented:
Testing this on my blogs. 
* jetpack_boost_minify_remove_stale_static_files() successfully removed a cache file once a dependent JS file was newer.
* The garbage collection works, once I fixed a bug in how it deletes files, and then checks them. Oops.
* The 404 tester works too, creating the legacy files when wp-content doesn't have mod_rewrite support, and then static files when I enabled it again. I toggled one of the concatenation modules to do this.

LiamSarsfield committed:
Remove unnecessary activate trait from minification classes

LiamSarsfield committed:
Add function to schedule 404 tester cronjob
